File: source/ts/src/core/tithi.ts
import { normalize360 } from '../utils/angle';
import { TITHI_SPAN } from '../utils/constants';
import type { TithiInfo } from '../types/elements';

/** Index 0-14 = Shukla Paksha, 15-29 = Krishna Paksha. */
export function computeTithiFromLongitudes(
  siderealMoon: number,
  siderealSun: number,
  name: string,
  paksha: string,
): TithiInfo {
  const angle = normalize360(siderealMoon - siderealSun);
  const index = Math.floor(angle / TITHI_SPAN);
  const elapsed = angle - index * TITHI_SPAN;
  const completionPercentage = (elapsed / TITHI_SPAN) * 100;
  const number = index < 15 ? index + 1 : index - 14;
  return {
    index,
    name,
    paksha,
    number,
    completionPercentage: Math.round(completionPercentage * 100) / 100,
    endTime: null,
  };
}

export function getTithiIndexAtTime(
  date: Date,
  getCachedMoon: (d: Date) => number,
  getCachedSun: (d: Date) => number,
): number {
  const angle = normalize360(getCachedMoon(date) - getCachedSun(date));
  return Math.floor(angle / TITHI_SPAN);
}

export function getTithiIndexFromLons(siderealMoon: number, siderealSun: number): number {
  return Math.floor(normalize360(siderealMoon - siderealSun) / TITHI_SPAN);
}

File: source/go/internal/core/tithi.go
package core

import (
	"math"

	"github.com/ishankgupta95/panchang/source/go/v5/internal/jsnum"
	"github.com/ishankgupta95/panchang/source/go/v5/internal/utils"
	"github.com/ishankgupta95/panchang/source/go/v5/types"
)

func ComputeTithiFromLongitudes(
	siderealMoon, siderealSun float64,
	name, paksha string,
) types.TithiInfo {
	angle := utils.Normalize360(siderealMoon - siderealSun)
	index := int(math.Floor(angle / utils.TithiSpan))
	elapsed := angle - float64(float64(index)*utils.TithiSpan)
	completionPercentage := (elapsed / utils.TithiSpan) * 100
	number := index - 14
	if index < 15 {
		number = index + 1
	}
	return types.TithiInfo{
		Index:                index,
		Name:                 name,
		Paksha:               paksha,
		Number:               number,
		CompletionPercentage: jsnum.Round(completionPercentage*100) / 100,
		EndTime:              nil,
	}
}

func GetTithiIndexAtTime(ms int64, getCachedMoon, getCachedSun LongitudeAt) int {
	angle := utils.Normalize360(getCachedMoon(ms) - getCachedSun(ms))
	return int(math.Floor(angle / utils.TithiSpan))
}

func GetTithiIndexFromLons(siderealMoon, siderealSun float64) int {
	return int(math.Floor(utils.Normalize360(siderealMoon-siderealSun) / utils.TithiSpan))
}
